Reintegrate aggregation support for OpenAPI

mbohlool opened the pull request:
Reintegrating changes of #46734

Changes summary:

- Extracted all OpenAPI specs to new repo `kube-openapi`
- Make OpenAPI spec aggregator to copy and rename any non-requal model (even with documentation change only).
- Load specs when adding APIServices and retry on failure until successful spec retrieval or a 404.
- Assumes all Specs except aggregator's Spec are static 
- A re-register of any APIService will result in updating the spec for that service (Suggestion for TPR: they should be registered to aggregator API Server, Open for discussion if any more changes needed for another PR.)

fixes #48548

mbohlool committed:
Revert "Aggregate OpenAPI specs"

This reverts commit 1a1d9a0394cbdb1d1e2412ae8f0157799eb5329c.

mbohlool committed:
Revert "Separate Build and Serving parts of OpenAPI spec handler"

This reverts commit 0a886ffaf8b9de97ef8134a4182b719ba2c6f22f.

sttts commented on staging/src/k8s.io/kube-openapi/README.md:
```diff
@@ -0,0 +1,14 @@
+# Kube OpenAPI
+
+This repo is the home for Kubernetes OpenAPI discovery spec generation. The goal 
+is to support a subset of OpenAPI features to satisfy kubernetes use-case but 
```
use-cases

mbohlool replied:
Done.

sttts commented on staging/src/k8s.io/kube-openapi/README.md:
```diff
@@ -0,0 +1,14 @@
+
+This repo is the home for Kubernetes OpenAPI discovery spec generation. The goal 
+is to support a subset of OpenAPI features to satisfy kubernetes use-case but 
+implement those subset with little to no assumption about the structure of the 
```
that subset

mbohlool replied:
done.

sttts commented on staging/src/k8s.io/kube-openapi/README.md:
```diff
@@ -0,0 +1,14 @@
+
+
+There are two main parts: 
+ - A model generator that goes through .go files, find types and generate model 
```
finds, generates

mbohlool replied:
should it be like this? `A model generator that goes through .go files, find and generate model definitions. `

sttts replied:
Am no native speaker, so I might be wrong: A model generator that goes through .go files, finds and generates model definitions.

sttts commented on staging/src/k8s.io/kube-openapi/cmd/openapi-gen/README:
```diff
@@ -0,0 +1,15 @@
+
+# OpenAPI Extensions
+OpenAPI spec can have extensions on types. To define one or more extensions on a type or its member
+add "+k8s:openapi-gen=x-kubernetes-$NAME:$VALUE" to the comment lines before type/member. A type/member can
```
single back quotes for verbatim, not `"`

sttts commented on staging/src/k8s.io/kube-openapi/cmd/openapi-gen/README:
```diff
@@ -0,0 +1,15 @@
+# OpenAPI Extensions
+OpenAPI spec can have extensions on types. To define one or more extensions on a type or its member
+add "+k8s:openapi-gen=x-kubernetes-$NAME:$VALUE" to the comment lines before type/member. A type/member can
+have multiple extensions. The rest of the line in the comment will be used as $VALUE so there is no need to
```
back-ticks around `$VALUE`

mbohlool replied:
other than back-ticks that whole `+k8s:open...` has?

sttts replied:
```
`+k8s:openapi-gen=x-kubernetes-$NAME:$VALUE`
```
No?

sttts commented on staging/src/k8s.io/kube-openapi/cmd/openapi-gen/README:
```diff
@@ -0,0 +1,15 @@
+OpenAPI spec can have extensions on types. To define one or more extensions on a type or its member
+add "+k8s:openapi-gen=x-kubernetes-$NAME:$VALUE" to the comment lines before type/member. A type/member can
+have multiple extensions. The rest of the line in the comment will be used as $VALUE so there is no need to
+escape or quote the value string. Extensions can be use to pass more information to client generators or
```
used

mbohlool replied:
done.

sttts commented on staging/src/k8s.io/kube-openapi/cmd/openapi-gen/README:
```diff
@@ -0,0 +1,15 @@
+add "+k8s:openapi-gen=x-kubernetes-$NAME:$VALUE" to the comment lines before type/member. A type/member can
+have multiple extensions. The rest of the line in the comment will be used as $VALUE so there is no need to
+escape or quote the value string. Extensions can be use to pass more information to client generators or
+documentation generators. For example a type my have a friendly name to be displayed in documentation or
```
s/my/might/

mbohlool replied:
done.

sttts commented on staging/src/k8s.io/kube-openapi/cmd/openapi-gen/generators/openapi.go:
```diff
@@ -0,0 +1,612 @@
+
+package generators
+
+import (
```
this code here is all old and only moved, isn't it?

mbohlool replied:
Yes, changes are restricted to necessary changes for the move (like imports).

sttts commented on staging/src/k8s.io/kube-openapi/pkg/handler/handler.go:
```diff
@@ -0,0 +1,188 @@
+	"github.com/googleapis/gnostic/OpenAPIv2"
+	"github.com/googleapis/gnostic/compiler"
+
+	genericmux "k8s.io/apiserver/pkg/server/mux"
```
this is kube import. Didn't we want to avoid this? This makes vendoring hard as it creates an import circle.

mbohlool replied:
Fixed it with an interface, also add import restrictions. thanks.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/handler/handler.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+// RegisterOpenAPIService registers a handler to provides access to provided swagger spec.
+func RegisterOpenAPIService(openapiSpec *spec.Swagger, servePath string, mux *genericmux.PathRecorderMux) (*OpenAPIService, error) {
```
a simple interface with `mux.Handle` as only method would be enough, wouldn't it?

mbohlool replied:
Done.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/handler/handler.go:
```diff
@@ -0,0 +1,188 @@
+// RegisterOpenAPIService registers a handler to provides access to provided swagger spec.
+func RegisterOpenAPIService(openapiSpec *spec.Swagger, servePath string, mux *genericmux.PathRecorderMux) (*OpenAPIService, error) {
+	if !strings.HasSuffix(servePath, JsonExt) {
+		return nil, fmt.Errorf("Serving path must ends with \"%s\".", JsonExt)
```
lower-case error. And `must end`. No full-stop.

mbohlool replied:
Applied both rules everywhere. Hopefully I didn't miss any. Will double check.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/handler/handler.go:
```diff
@@ -0,0 +1,188 @@
+
+// RegisterOpenAPIService registers a handler to provides access to provided swagger spec.
+func RegisterOpenAPIService(openapiSpec *spec.Swagger, servePath string, mux *genericmux.PathRecorderMux) (*OpenAPIService, error) {
+	if !strings.HasSuffix(servePath, JsonExt) {
```
This pre-conditions belongs into the godoc.

mbohlool replied:
Done.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/handler/handler.go:
```diff
@@ -0,0 +1,188 @@
+		getDataAndETag := file.getDataAndETag
+		mux.Handle(path, gziphandler.GzipHandler(http.HandlerFunc(
+			func(w http.ResponseWriter, r *http.Request) {
+				if r.URL.Path != path {
```
why this comparison? Shouldn't the mux do the not-found logic?

mbohlool replied:
I was there in other handlers, I thought there should be an unknown (to me) reason for this. I agree it is not necessary.

sttts replied:
@deads2k anything like this needed in the path-muxed handlers? I think you rewrote it recently.

mbohlool replied:
I don't see it in other places anymore with a quick search. I removed it. Please let me know if I need to add it back.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/handler/handler.go:
```diff
@@ -0,0 +1,188 @@
+				w.Header().Set("Etag", etag)
+
+				// ServeContent will take care of caching using eTag.
+				http.ServeContent(w, r, path, o.lastModified, bytes.NewReader(data))
```
we don't want to keep the lock while serving the content. Copy the o.lastModified and unlock after that and `getDataAndETag()`

mbohlool replied:
I need to copy data too as it may change while we serve it.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/handler/handler.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+func (o *OpenAPIService) getSwaggerBytes() ([]byte, string) {
+	return o.specBytes, o.specBytesETag
```
this requires the lock. This belongs into a godoc. Same below.

mbohlool replied:
I will revise locking with a RW lock in this code (and also make the change atomic).

sttts commented on staging/src/k8s.io/kube-openapi/pkg/handler/handler.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+func (o *OpenAPIService) UpdateSpec(openapiSpec *spec.Swagger) (err error) {
+	defer o.mu.Unlock()
```
A RW lock would be more appropriate.

mbohlool replied:
ack

sttts commented on staging/src/k8s.io/kube-openapi/pkg/handler/handler.go:
```diff
@@ -0,0 +1,188 @@
+	if err != nil {
+		return err
+	}
+	o.specPb, err = toProtoBinary(o.specBytes)
```
all these lines will render the object invalid because later accesses to the nil members will possibly panic. Would be better to make the whole modification atomic.

mbohlool replied:
ack

sttts commented on staging/src/k8s.io/kube-openapi/pkg/handler/handler.go:
```diff
@@ -0,0 +1,188 @@
+)
+
+const (
+	JsonExt = ".json"
```
can all these constant be private?

mbohlool replied:
yes

sttts commented on staging/src/k8s.io/kube-openapi/pkg/handler/handler.go:
```diff
@@ -0,0 +1,188 @@
+	MimePbGz = "application/x-gzip"
+)
+
+type OpenAPIService struct {
```
godoc

sttts commented on staging/src/k8s.io/kube-openapi/pkg/handler/handler.go:
```diff
@@ -0,0 +1,188 @@
+type OpenAPIService struct {
+	orgSpec      *spec.Swagger
+	lastModified time.Time
+	mu           sync.Mutex
```
what does the mutex protect? godoc would be nice.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -0,0 +1,480 @@
+
+	"github.com/go-openapi/spec"
+
+	"k8s.io/apimachinery/pkg/conversion"
```
again kube imports. Didn't we want to avoid them?

mbohlool replied:
Fixed.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -0,0 +1,480 @@
+	return true
+}
+
+func equalSchemaArray(s1, s2 []spec.Schema) bool {
```
For two schemas x and y, the following slices are equal: `[x,x,y]` and `[x,y,y]`.

But, `[x,y]`, `[x,y,y]` are not. Also `nil` and `[]` are not equal.

Intentional?

mbohlool replied:
Generally we don't have duplicates in our usecase, but it is a good idea to check for them.

mbohlool replied:
I think the method is on the safe side and may return false positive that will result in duplicating an schema that we could reuse. Because we don't have those edge cases in our schema and the repo statement is specific to kubernetes use-cases, I think a TODO should be fine for now.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -0,0 +1,480 @@
+	if !equalSchemaArray(s1.AnyOf, s2.AnyOf) {
+		return false
+	}
+	if !equalSchemaArray(s1.OneOf, s2.OneOf) {
```
here the comparison gives the wrong result. Only one may be true, but `{oneOf: [x,x]}` is never fulfilled while `{oneOf: [x]}` could. Still both are considered equal.

mbohlool replied:
Sorry, I am not sure if I understand this comment. Do you mean `{oneOf: [x]}` and `{oneOf: [x,x]}` should be considered equal?

sttts replied:
I oversimplified the example. Now `{oneOf: [x,y,y]}` and `{oneOf: [x,x,y]}` are considered equal, aren't they? For a matching `x`. the first will be valid, the second won't.

mbohlool replied:
Is this the same issue with equalSchemaArray above? This is the part I don't get:

> the first will be valid, the second won't.

You have two oneof examples, and the line of code compares if they are equal and with current code it assumes they are (and they should not be equal). But why one of them is valid the other is not?

sttts replied:
With `valid` I meant that there is a schema of the shape `s={oneOf: [x,y,y]}` such that a JSON object `j` exists where `s'={oneOf: [x,x,y]}` does not match. So `s` and `s'` should be equal.

Example: `s={oneOf: [{type: integer}, {type: string}, {type: string}]}`. Then `j=42` matches. At the same time `j` does not match `s={oneOf: [{type: integer}, {type: integer}, {type: string}]}`.

Does that make sense?
 there is a JSON object which matches the a schema of the shape

mbohlool replied:
Sorry, still no :(

`j` is an integer and in both cases {type: integer} is one of possibilities and it should match both cases. unless order means something in `oneOf`.

mbohlool replied:
I see now. I mistaken `OneOf` with `AnyOf`. Duplication in `OneOf` doesn't make sense from design point of view because that duplicate can never match with anything. Do you think we should have an special case for this or just prevent duplication in `OneOf`?

sttts replied:
duplication doesn't make sense in most cases, but it is totally valid from a user point of view and JSON-Schema.

Coming back to my original question: an array is sorted in OpenAPI. Why do we try to unify anything and not just compare index by index?

sttts commented on staging/src/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -0,0 +1,480 @@
+	if !equalStringArray(s1.Required, s2.Required) {
+		return false
+	}
+	return len(s1.Enum) == 0 && len(s2.Enum) == 0 && len(s1.Dependencies) == 0 && len(s2.Dependencies) == 0
```
So `EqualSchema(x, x)` returns falls if `Enum` is set?

mbohlool replied:
Yes, added to godoc.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -0,0 +1,480 @@
+	}
+	for k, v := range source.Paths.Paths {
+		if _, found := dest.Paths.Paths[k]; found {
+			return fmt.Errorf("Unable to merge: Duplicated path %s", k)
```
lower-case errors

mbohlool replied:
Applied it everywhere in kube-openapi.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -0,0 +1,480 @@
+	renames := []Rename{}
+	for k, v := range source.Definitions {
+		v2, found := dest.Definitions[k]
+		if found || usedNames[k] {
```
doesn't `found` imply `usedNames[k]` due to the `usedNames` loop 10 lines above?

mbohlool replied:
no, it is the other way around. `usedNames[k]` imply `found`. A name can be used but not in `dest` (the names in rename list are also added to used during this loop). I will simplify the if.

sttts replied:
my imply direction is right I think. If it is in dest, it is used. It's what you are saying as well.

mbohlool replied:
if `found==True`, `usedNames[k]` is `True`.
if `usedNames[k]==True`, then `found` may not be `True`.

I guess you're right about `imply`. Anyway, checking only for `usedNames[k]` should be sufficient here.

sttts replied:
👍

sttts commented on staging/src/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -0,0 +1,480 @@
+		}
+	}
+	for _, r := range renames {
+		renameDefinition(source, r.from, r.to)
```
In the godoc you promised me not to modify `source`. But now you do? 😱

sttts replied:
And you never check whether `r.to` is used in `source`. If it is, this rename override the old value for `r.to`. 😱

mbohlool replied:
I cloned source, so technically I am not changing "your" source. Will rename the variable to prevent any confusion.

mbohlool replied:
The logic with `usedNames` above will make sure r.to is not used in source (we iterate through source, and rename to names that are not used and add these new names to `usedNames` so we don't use them again).

sttts replied:
right @ clone. Missed that line.

sttts replied:
about `r.to`. To make sure that `r.to` is **fresh for dest** (because userNames are those of dest initialy). So a later `k` value in the `for k, v := range source.Definitions {` loop might be conflict with an earlier chosen `newName`, can't it? E.g. `dest={"foo":1}` and `source={"foo":1, "foo_v2":2}`. The loop will consider `k="foo"`. It will notice `"foo"` is not fresh and will choose `newName="foo_v2"`. It looks fresh for `dest` and a `Rename{from: "foo", to: "foo_v2"}` is created. The next loop iteration will consider `k="foo_v2"`. `"foo_v2"` is also not fresh now, so a rename to `"foo_v2_v2"` is created. At the end, the renames are applied, giving `source={"foo_v2":1,"foo_v2_v2":1}`, but `source={"foo_v2":1,"foo_v2_v2":2}` would be correct.

mbohlool replied:
In your example, source should be `source={"foo":3, "foo_v2":2}` because `"foo":1` is the same definition and no rename will happen for it. But with this new source shows the bug. I think if we add all source names to `usedNames` list, it should fix the problem. I need to think a little more about it to make sure it won't create another bug. What do you think?

mbohlool replied:
well adding them to usedNames doesn't help. I will fix it another way. Will keep you updated.

mbohlool replied:
I've fixed this by adding another check to make sure we don't have the name in source before adding the rename job.

sttts replied:
yeah, the last comment's solution sounds good.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -0,0 +1,480 @@
+
+// Clone OpenAPI spec
+func CloneSpec(source *spec.Swagger) (*spec.Swagger, error) {
+	if ret, err := cloner.DeepCopy(source); err != nil {
```
we better have a test for this deepcopy.

sttts replied:
Can there be `github.com/go-openapi/jsonreference.Ref` instances in a swagger spec? If yes, a `Ref` consists of private variables. Our reflection based deepcopy won't see them I bet. I hope I am wrong. If not, the clone is corrupt.

mbohlool replied:
Yes, It was working with references for sure, but there is also the problem of importing cloner, I will use json marshal/unmarshal for now and will add a TODO to find a better way to do this.

sttts replied:
json marshaling is fine. I am doing the same in my pending static deepcopy PR, for this very code line.

sttts commented:
@mbohlool can you add import restrictions to `hack/verify-staging-imports.sh`? We have imports now from apiserver and apimachinery which we tried to avoid.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -0,0 +1,480 @@
+)
+
+const (
+	DefinitionPrefix = "#/definitions/"
```
this should be private

mbohlool replied:
Done.

mbohlool commented:
@sttts this is ready for another round of review. The only thing I added a TODO for was `equalSchemaArray` function. It is not clear what is the right way to do it and specially in our case we don't have duplicates items in our schema array.

mbohlool commented:
I gave it a try to implement an `equalSchemaArray` and `equalStringArray`. It is a separate commit.

cblecker commented:
cross-build is currently broken: https://github.com/kubernetes/kubernetes/issues/48887

sttts commented on staging/src/k8s.io/kube-openapi/cmd/openapi-gen/generators/openapi.go:
```diff
@@ -577,7 +577,7 @@ func (g openAPITypeWriter) generateMapProperty(t *types.Type) error {
 	switch elemType.Kind {
 	case types.Builtin:
-		return fmt.Errorf("please add type %v to getOpenAPITypeFormat function.", elemType)
+		return fmt.Errorf("please add type %v to getOpenAPITypeFormat function", elemType)
```
funny error message. Never seen "please do something" in an error ;)

sttts commented on staging/src/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -213,10 +209,29 @@ func equalSchemaMap(s1, s2 map[string]spec.Schema) bool {
+}
+
+func equalSchemaArray(in1, in2 []spec.Schema) bool {
+	s1 := toSchemaArrayWithUniqueItems(in1)
```
compare my `oneOf` example. At least for that construct we must not unify. I guess for all other JSON-Schema constructs that's fine.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -273,7 +288,29 @@ func equalSchemaOrArray(s1, s2 *spec.SchemaOrArray) bool {
+	}
+	for _, v1 := range in {
+		found := false
+		for _, v2 := range out {
```
sort + unification would be simpler and faster

mbohlool replied:
Yes, that is better for string arrays. Done.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -320,6 +357,8 @@ func equalIntPointer(s1, s2 *int64) bool {
 
 // EqualSchema returns true if models have the same properties and references
 // even if they have different documentation.
+// Note: Dependencies and Enum are not supported. EqualSchema will return
```
any reason what we do this? why not implement them?

mbohlool replied:
I don't expect we use `Dependencies`. For `Enum` it is an array of interfaces. I am not sure how to compare them. Anyway, I don't expect to use that either any time soon and also not implementing them is not terrible. It is a False positive that will result in a unnecessary schema duplication. I can spend time figure out how to implement them, but I don't think it worth it.

sttts replied:
I guess we have a limited number of a types for enums. It's just odd to have this logic in an equality implementation and not in a validation somewhere else.

mbohlool replied:
I can implement enums with DeepEqual I guess.

mbohlool replied:
Decided to get rid of custom equal methods and just use `reflect.DeepEqual`.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -474,9 +514,15 @@ func MergeSpecs(dest, source *spec.Swagger) error {
 // Clone OpenAPI spec
 func CloneSpec(source *spec.Swagger) (*spec.Swagger, error) {
-	if ret, err := cloner.DeepCopy(source); err != nil {
+	// TODO(mehdy): Find a faster way to clone an spec
```
how often do we clone a spec? Once per APIService change? That's probably not often.

mbohlool replied:
Yes, but it still feels terrible to clone with marshalling/unmarshalling. It is painfully slow (~500ms). I think we can keep the TODO and when we got a better deepcopy, we can use it here.

sttts replied:
Really half a second? How big is our swagger doc?

mbohlool replied:
Well big :) Look here: https://docs.google.com/document/d/1EcnvLRqP9-5mqQ-kghfnB3GtB8eIrkVQjn1B16NRrr0

sttts commented on staging/src/k8s.io/kube-openapi/pkg/common/common.go:
```diff
@@ -41,6 +43,10 @@ type OpenAPIDefinitionGetter interface {
 	OpenAPIDefinition() *OpenAPIDefinition
 }
 
+type Handler interface {
```
would prefer PathHandler

mbohlool replied:
Done.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/handler/handler.go:
```diff
@@ -0,0 +1,203 @@
+	orgSpec      *spec.Swagger
+	lastModified time.Time
+
+	// mu protects All members of this service.
```
nit: move it to the top of the struct.

mbohlool replied:
Done.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/handler/handler.go:
```diff
@@ -0,0 +1,203 @@
+		getDataAndETag := file.getDataAndETag
+		handler.Handle(path, gziphandler.GzipHandler(http.HandlerFunc(
+			func(w http.ResponseWriter, r *http.Request) {
+
```
nit: no empty line

mbohlool replied:
Fixed.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/handler/handler.go:
```diff
@@ -0,0 +1,203 @@
+			func(w http.ResponseWriter, r *http.Request) {
+
+				data, etag, lastModified := getDataAndETag()
+				w.Header().Set("Etag", etag)
```
Don't we need an if-clause for etags? I.e. if the etag head from the client matches ours, return early without http body?

sttts replied:
Oh, ServeContent does that 👍

sttts commented on staging/src/k8s.io/kube-openapi/pkg/handler/handler.go:
```diff
@@ -0,0 +1,203 @@
+}
+
+func (o *OpenAPIService) getSwaggerBytes() ([]byte, string, time.Time) {
+	defer o.rwMutex.RUnlock()
```
nit: usually the style is: `lock(); defer unlock()`

sttts replied:
ratio: if the lock panics for any reason the unlock won't be called.

mbohlool replied:
Fixed.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -443,15 +482,16 @@ func MergeSpecs(dest, source *spec.Swagger) error {
 			newName := fmt.Sprintf("%s_v%d", k, i)
-			for usedNames[newName] {
+			_, foundInSource := sourceCopy.Definitions[newName]
+			for usedNames[newName] || foundInSource {
```
you have to update `foundInSource`. Otherwise, this loops infinitely.

mbohlool replied:
Ops. Fixed.

sttts commented:
A general question: the reason to implement equality manually is that `reflect.DeepEqual` is too slow? Or do we really need the independence of the order or arrays?

mbohlool commented:
`reflect.DeepEqual` needs exact equal models. Main reason was models with different documentation should not be duplicated. Order of the maps/arrays is also another reason though one can argue that we should duplicate schemas if they have different orders. While I fix your comments above, I will think about this.

cblecker commented:
/test pull-kubernetes-cross

mbohlool commented:
@sttts @lavalamp @deads2k 
When should we duplicate schemas?
The question of what we considered to be "Equal" schema came out on this PR and I discussed it in some detail with @pwittrock too. For example, do we consider two schema the same if the only difference is documentation? or do we consider them the same if they have different order of same fields?

This depends on use case. For clients, for example, the order is important (it goes into the order of parameters in API function calls). For some other tools like auto-complete, the order does not matter, the documentation may. Because of these, I am leaning toward copying a Model if there is any difference (even documentation or order of fields). What do you guys think?

Another problem is the naming of duplicates. Right now we add _v1, _v2, etc suffix to the name of the new models. That may imply some kind of order that we don't intend. We can go with adding source of these models to the name (e.g. PodTemplateForExampleApiServer) but that could get inconsistent if two APIService share the same model (the name of the model would be choiced randomly by the order of aggregation). I think we should just name them `*_copy1`, `*_copy2`, etc. There would be inconsistent naming because of the order, but we can't prevent inconsistency in general. Any ideas here?

mbohlool commented:
@sttts After discussing this with @lavalamp and thinking more about it, I decided to just go with `reflect.DeepEqual` and get rid of my custom EqualSchema functions. We can add them back later if we see it fit. I think I also addressed other comments, Please take another look.

mbohlool commented:
@sttts Friendly ping. Is `kube-openapi` part good to go into its own repo?

sttts commented on staging/src/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -0,0 +1,277 @@
+
+	"github.com/go-openapi/spec"
+
+	"reflect"
```
import order

sttts commented on staging/src/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -0,0 +1,277 @@
+	}
+	renames := []Rename{}
+	for k, v := range sourceCopy.Definitions {
+		v2, found := dest.Definitions[k]
```
nit: move this line into the if clause

sttts commented on staging/src/k8s.io/kube-openapi/pkg/builder/doc.go:
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 */
 
-// Package openapi contains code to generate OpenAPI discovery spec (which
+// Package openapi/builder contains code to generate OpenAPI discovery spec (which
```
nit: Package builder contains....

sttts commented on staging/src/k8s.io/kube-openapi/pkg/common/doc.go:
```diff
@@ -14,5 +14,6 @@ See the License for the specific language governing permissions and
 
-// package openapi holds shared codes and types between open API code generator and spec generator.
-package openapi
+// package openapi/common holds shared codes and types between open API code
```
nit: Package common ....

sttts commented on staging/src/k8s.io/kube-openapi/pkg/handler/handler.go:
```diff
@@ -0,0 +1,202 @@
+	return fmt.Sprintf("\"%X\"", sha512.Sum512(data))
+}
+
+// BuildAndRegisterOpenAPIService first build and spec and then registers a handler to provides access to it.
```
builds the spec

sttts commented on staging/src/k8s.io/kube-openapi/pkg/handler/handler.go:
```diff
@@ -0,0 +1,202 @@
+// json file and will also serve .pb and .gz files.
+func RegisterOpenAPIService(openapiSpec *spec.Swagger, servePath string, handler common.PathHandler) (*OpenAPIService, error) {
+	if !strings.HasSuffix(servePath, jsonExt) {
+		return nil, fmt.Errorf("serving path must ends with \"%s\"", jsonExt)
```
must end with

sttts commented on staging/src/k8s.io/kube-openapi/pkg/handler/handler.go:
```diff
@@ -0,0 +1,202 @@
+		return nil, err
+	}
+
+	mime.AddExtensionType(".json", mimeJson)
```
Would expect this in an init func.

sttts commented on staging/src/k8s.io/kube-openapi/pkg/util/trie.go:
```diff
@@ -0,0 +1,79 @@
+
+package util
+
+// A simple trie implementation with Add an HasPrefix methods only.
```
and

sttts commented on staging/src/k8s.io/kube-openapi/pkg/util/trie.go:
```diff
@@ -0,0 +1,79 @@
+	word     string
+}
+
+// New creates a Trie and add all strings in the provided list to it.
```
NewTrie

sttts commented:
@mbohlool couple of nits. Otherwise, it looks good now 👍

mbohlool committed:
Aggregate OpenAPI spec

mbohlool commented:
@sttts I addressed your comments. I will now rebase and push these changes to their repo.

mbohlool committed:
Update OpenAPI spec

mbohlool committed:
Update main repo references to new kube-openapi repo

mbohlool committed:
Update Bazel

mbohlool committed:
Update Godeps to use kube-openapi

mbohlool commented:
@sttts I've moved kube-openapi files in their repo and added godep here. The only thing I changed was moving generators from "/cmd" to "/pkg" and kept the main file of generator in the main repo (we cannot import it with godep as it is a runnable application).

I also pushed the final commit and it is ready to review. I will work on any check fails meanwhile. 

Please take another look.

cheftako commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:
```diff
@@ -322,24 +291,12 @@ func (s *APIAggregator) AddAPIService(apiService *apiregistration.APIService) {
 // RemoveAPIService removes the APIService from being handled.  It is not thread-safe, so only call it on one thread at a time please.
 // It's a slow moving API, so its ok to run the controller on a single thread.
-func (s *APIAggregator) RemoveAPIService(apiServiceName string) {
+func (s *APIAggregator) RemoveAPIService(apiServiceName string) error {
```
Why? Looks like we are only returning nil right now. Till we actually return an error, why do this. In addition I can be convinced that its ok to return an error adding an API when there is an OpenAPI error. However I do not think its ok to not locally remove an API.

mbohlool replied:
I just added this to be consistent with `AddAPIService`. But it make sense. I removed this.

cheftako commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -0,0 +1,252 @@
+}
+
+func (r *inMemoryResponseWriter) Header() http.Header {
+	if r.header == nil {
```
Do we need to worry about access contention here? Would we better off just using a constructor which insures this exists?

mbohlool replied:
Done.

cheftako commented on vendor/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -0,0 +1,276 @@
+	}
+}
+
+func renameDefinition(s *spec.Swagger, old, new string) {
```
"new" is not a reserved word in go but it does have special meaning. Might be worth changing it slightly, especially as even github's syntax highlighting flags it as special.

mbohlool replied:
Sure, that needs another PR to kube-openapi repo.

cheftako commented on vendor/k8s.io/kube-openapi/pkg/util/trie.go:
```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Is this now duplicated code? I did not see you deleting the copy in staging/src/k8s.io/apiserver/pkg/util/trie/trie.go

mbohlool replied:
Yes, it is duplicate. It is a simple utility that used in both apiserver and here. I cannot import it from API server as it creates a dependency cycle. The only other option is to create a util repo and move it there, that would be overkill for this simple utility but when we had that repo for other utilities, we can move this there and remove duplicates.

mbohlool replied:
Fixed.

mbohlool commented:
@cheftako address your comments. PTAL.

sttts commented:
@mbohlool let me know when it's a good time for another review

mbohlool commented:
@sttts It is ready to review.

> I also pushed the final commit and it is ready to review. I will work on any check fails meanwhile.

> Note: This PR needs constant rebase and tests will fail because of that. I will rebase/squash after LGTM/Approval)

sttts commented:
@mbohlool sorry for no answer, am travelling. Will take a look tomorrow.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/openapi/openapi.go:
```diff
@@ -30,10 +30,10 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
-	"k8s.io/apiserver/pkg/util/trie"
+	"k8s.io/kube-openapi/pkg/util"
```
having the trie code there is odd. Why isn't that in k8s.io/util?

deads2k replied:
> having the trie code there is odd. Why isn't that in k8s.io/util?

Never used anywhere else.  This was a one-use util

mbohlool replied:
I has been used in apiserver and kube-openapi. Does that mean I need to move it to k8s.io/util?

sttts replied:
It feels odd to me to share utils between repos if they are not in k8s.io/util. But I don't feel strongly about this one.

mbohlool replied:
k8s.io/utils is a new form repo and for this instance I actually think we can move it there. Just for the sake of getting this PR in faster, I suggest we do it in another PR.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:
```diff
@@ -176,6 +144,12 @@ func (c *Config) SkipComplete() completedConfig {
 
 // New returns a new instance of APIAggregator from the given config.
 func (c completedConfig) NewWithDelegate(delegationTarget genericapiserver.DelegationTarget) (*APIAggregator, error) {
+
```
nit: no new line

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:
```diff
@@ -292,21 +261,21 @@ func (s *APIAggregator) AddAPIService(apiService *apiregistration.APIService) {
 	}
 	proxyHandler.updateAPIService(apiService)
+	if err := s.openAPIAggregator.loadApiServiceSpec(proxyHandler, apiService); err != nil {
+		return err
```
what are the consequences of such an error? If I see it correctly, we keep the old handler installed? Will the aggregator still be able to serve the old openapi spec?

mbohlool replied:
Unless the call succeed, neither the handler nor its state would change (at least that was my intention). Returning error here will force the controller to retry later.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -0,0 +1,253 @@
+
+	"github.com/emicklei/go-restful"
+	"github.com/go-openapi/spec"
+	"k8s.io/apiserver/pkg/authentication/user"
```
empty line above between kube and non-kube imports

mbohlool replied:
I wasn't sure what is the convention here. My understanding from some other files was we group in three sections. First is core packages, second are dependencies and third is internal references (references in the same repo). Is there any document for this convention? I am curious to understand it. (I will fix this anyway).

sttts replied:
I fear it's a bit mixed. Most code probably has "stdlib, 3rdparty, kube". Some of the staging repos (mostly by @deads2k I think) have "stdlib, 3rdparty, kube, that very staging repo". Both go from near-to-standard-go to concrete. Your order was mostly that of david, but the apiserver package in both blocks was misplaced. Anyway, it's only cosmetics :)

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -0,0 +1,253 @@
+	"k8s.io/kube-openapi/pkg/builder"
+	"k8s.io/kube-openapi/pkg/common"
+	"k8s.io/kube-openapi/pkg/handler"
+
```
no empty line here

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -0,0 +1,253 @@
+)
+
+const (
+	aggregatorUser = "system:aggregator"
```
don't we have this constant anywhere else?

sttts replied:
or is only for openapi?!

mbohlool replied:
Right now OpenAPI is the only component of aggregator that make a call that is not initiated by user (hence does not have callers information), but if we end up doing those calls for any other reason, we should share one user.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -0,0 +1,253 @@
+	openAPIService *handler.OpenAPIService
+
+	// Aggregator's OpenAPI spec.
+	aggregatorOpenAPISpec *spec.Swagger
```
aggregated or aggregator (= for the aggregator api group)? please carify name and/or comment.

mbohlool replied:
aggregator (i.e. for the aggregator api group). I will clarify it in the comment.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -0,0 +1,253 @@
+	aggregatorOpenAPISpec *spec.Swagger
+
+	// Local delegate's OpenAPI spec.
+	localDelegatesOpenAPISpec *spec.Swagger
```
what is the **local** delegate?

mbohlool replied:
The main API server (the one that has most of specs). It called localDelegate other places I think. Do you have a better name suggestion?

sttts replied:
I thought to have seen "delegate" only in the context of the apiserver delegation, i.e. in-process. So the prefix looked superfluous. Do we use that also to talk about aggregated servers? Maybe in-process delegates is better as a name? Not sure.

mbohlool replied:
Renamed it to `inProcess...`. I like is as much as the previous one (which is not much :D)

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -0,0 +1,253 @@
+}
+
+// Load openAPI spec from /swagger.json of given handler.
+func (s *openAPIAggregator) loadOpenAPISpec(handler http.Handler) (*spec.Swagger, error) {
```
"load" usually implies side-effects. Are there any of this func? If not, better call it "handlerOpenAPISpec". If yes, please specify which.

mbohlool replied:
Handler means it gets an input and when it got it will return a result. Method loads spec and return it. I am not sure if load means side effect, but I can rename it to "get" or "retrieve" or "download"?

sttts replied:
all three seem to be clearer.

mbohlool replied:
does your comment here applies to `loadApiServiceSpec` method too? That one has a side effect and I guess no.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -0,0 +1,253 @@
+	}
+	// Remove any non-API endpoints from aggregator's spec. aggregatorOpenAPISpec
+	// is the source of truth for all non-api endpoints.
+	aggregator.FilterSpecByPaths(s.aggregatorOpenAPISpec, []string{"/apis/"})
```
I guess the `/apis` discovery endpoint is not filtered out?

mbohlool replied:
It is filtered out. The aggregation will fail if two of these has `/apis` discovery point.

mbohlool replied:
That is true. It is not filtered out. The reason is the aggregator is responsible for that discovery endpoint and we should let it provide spec for the endpoint.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -0,0 +1,253 @@
+	contextMapper request.RequestContextMapper
+}
+
+func newOpenAPIAggregator(delegateHandler http.Handler, webServices []*restful.WebService, config *common.Config, pathHandler common.PathHandler, contextMapper request.RequestContextMapper) (s *openAPIAggregator, err error) {
```
`newSomething` is supposed to be a side-effect free constructor. This func is not side-effect free, but installs into `pathHandler`. Better call it `buildAndRegister...` or something like that.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -0,0 +1,253 @@
+	if a[i].priority != a[j].priority {
+		// Sort by priority, higher first
+		return a[i].priority > a[j].priority
+	} else {
```
no else clause necessary. Just call return directly. That's more golang-idiomatic.

sttts commented on vendor/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -0,0 +1,276 @@
+}
+
+func (s *referenceWalker) walkRef(ref spec.Ref) spec.Ref {
+	if ref.String() != "" {
```
```
if refStr := ref.String(); refStr != "" {
```

sttts commented on vendor/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -0,0 +1,276 @@
+		// inside the same spec file. If that is the case, walk through
+		// those definitions too.
+		// We do not support external references yet.
+		if strings.HasPrefix(refStr, definitionPrefix) {
```
`strings.HasPrefix(refStr, definitionPrefix)` implies `ref.String() != ""`. No need for the first if-clause at all.

sttts commented on vendor/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -0,0 +1,276 @@
+		// We do not support external references yet.
+		if strings.HasPrefix(refStr, definitionPrefix) {
+			def := s.root.Definitions[refStr[len(definitionPrefix):]]
+			s.walkSchema(&def)
```
can this loop for circular references? Do we validate this somewhere?

sttts replied:
AFAIK circular JSON-Schemas are not illegal, e.g. to express algebraic data structures.

sttts commented on vendor/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -0,0 +1,276 @@
+	}
+}
+
+func (s *referenceWalker) Start() {
```
as an idea to test this code: use the fuzzer to create random schemas, use a reflection based walker to find all `spec.Schema` structs, call your walker and compare which schemas you see with the other list.

mbohlool replied:
Do you want that test to be added before this PR merged? I am sending another PR to kube-openapi repo to address your comments, but I don't think we need to keep this PR blocked on that.

sttts commented on vendor/k8s.io/kube-openapi/pkg/util/trie.go:
```diff
@@ -14,17 +14,17 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package trie
```
without any import this should definitely go to k8s.io/util. Here – used by other components – it is not at the right place.

mbohlool replied:
Will do this in my other PR.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -0,0 +1,253 @@
+)
+
+type openAPIAggregator struct {
+	// Map of API Services' OpenAPI specs by their name
```
what exactly is this name? An api group?

mbohlool replied:
this is the name user provided for api service (`ApiService.ObjectMeta.Name`).

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -0,0 +1,253 @@
+	s.openAPISpecs[apiService.Name] = &openAPISpecInfo{
+		serviceName: apiService.Name,
+		spec:        openApiSpec,
+		priority:    max(apiService.Spec.VersionPriority, apiService.Spec.GroupPriorityMinimum),
```
these two values have a totally independent scale. A max does not make sense.

mbohlool replied:
My understanding of these two values was completely wrong. I've changed sorting algorithm to fix this.

mbohlool replied:
Also added a test case for it.

sttts commented on vendor/k8s.io/kube-openapi/pkg/common/doc.go:
```diff
@@ -14,5 +14,6 @@ See the License for the specific language governing permissions and
 
-// package openapi holds shared codes and types between open API code generator and spec generator.
-package openapi
+// package common holds shared codes and types between open API code
```
nit: code

sttts commented on vendor/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -0,0 +1,276 @@
+	}
+	for k, v := range sourceCopy.Paths.Paths {
+		if _, found := dest.Paths.Paths[k]; found {
+			return fmt.Errorf("unable to merge: Duplicated path %s", k)
```
nit: lowercase "duplicated"

sttts commented on vendor/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -0,0 +1,276 @@
+			renames = append(renames, Rename{from: k, to: newName})
+			usedNames[newName] = true
+		} else {
+			usedNames[k] = true
```
no need for this case

sttts commented on vendor/k8s.io/kube-openapi/pkg/aggregator/aggregator.go:
```diff
@@ -0,0 +1,276 @@
+	// First remove unwanted paths
+	prefixes := util.NewTrie(keepPathPrefixes)
+	orgPaths := sp.Paths
+	if orgPaths == nil {
```
`len(orgPaths) == 0`

mbohlool replied:
orgPaths is not an array and does not have len. Surprised? I know :)

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -0,0 +1,253 @@
+	handler = s.handlerWithUser(handler, &user.DefaultInfo{Name: aggregatorUser})
+	handler = request.WithRequestContext(handler, s.contextMapper)
+
+	req, err := http.NewRequest("GET", "/swagger.json", nil)
```
should we have some context here with a timeout? How does this behave?

mbohlool replied:
I think `http.TimeoutHandler` will do the trick here.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -0,0 +1,253 @@
+		}
+		return openApiSpec, nil
+	default:
+		return nil, fmt.Errorf("failed to retrive openAPI spec, http error: %s", writer.String())
```
What about the http code? Would print that here, not only the body.

sttts replied:
oh ic, the String() func does this.

mbohlool commented:
@sttts I addressed all of your comments. I sent PR https://github.com/kubernetes/kube-openapi/pull/1 to address your comments in `kube-openapi` repo.

Please take another look. Thanks.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -95,40 +96,62 @@ func newOpenAPIAggregator(delegateHandler http.Handler, webServices []*restful.W
 	} else {
-		// Sort by service name.
-		return a[i].serviceName < a[j].serviceName
+		iPriority = a.groupPriorities[a.specs[i].apiService.Spec.Group]
```
I fear our logic there is more complicated. Look for some quiet moment and read what we came up with shortly before 1.7 release: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/v1beta1/types.go#L59 ;-) Wondering, can we expose that logic from kube-aggregator code somehow and reuse it here?

mbohlool replied:
`groupPriorities` map is created to specifically tackle that. before sort, I am getting a max on `GroupPriorityMininum` for all ApiSevices with the same group name and use that priority here. As far as I can tell from the comment, this should work.

sttts replied:
Oh, alright then. Didn't realize that it's not the priority of each group version directly :+1:

sttts commented:
Much better now! 👍  The priority question is still open, compare my comment.

sttts commented:
Then I think after a rebase it's ready to go 🚀

mbohlool commented:
This is now blocked on https://github.com/kubernetes/k8s.io/pull/68

lavalamp commented:
/approve no-issue

mbohlool commented:
/test pull-kubernetes-verify

mbohlool commented:
@sttts I rebased it and now fixing any failure tests (that mostly is because of godep). Do you want me to self LGTM it after the process (obviously if there is no real code-change)?

sttts commented on staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go:
```diff
@@ -18,30 +18,22 @@ package routes
-	}
-	service, err := apiserveropenapi.RegisterOpenAPIService(openapiSpec, "/swagger.json", mux)
+func (oa OpenAPI) Install(c *restful.Container, mux *mux.PathRecorderMux) {
+	_, err := handler.BuildAndRegisterOpenAPIService("/swagger.json", c.RegisteredWebServices(), oa.Config, mux)
```
`c` is only for one of the servers in a delegation chain. What about the others? They also have webservices we have to consider (btw, the same for swagger).

sttts replied:
Compare cmd/kube-apiserver/app/server.go:141. If you want to test that the chain is properly considered in the openapi setup, `cmd/kube-apiserver/app/testing/server_test.go` might be a good place to add a test.

mbohlool replied:
The same as the other comment?

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:
```diff
@@ -249,31 +219,29 @@ func (c completedConfig) NewWithDelegate(delegationTarget genericapiserver.Deleg
-		if err := s.updateOpenAPISpec(); err != nil {
+	if openApiConfig != nil {
+		s.openAPIAggregator, err = buildAndRegisterOpenAPIAggregator(
+			s.delegateHandler,
```
What about the delegate of the delegate? The chain is not necesserily depth 1. Already today we have the apiextensions server behind the main kube-apiserver.

mbohlool replied:
I added an issue for that. I think we should CRDs need to register themselves with aggregator api server. Whatever we decide, though, I think it should be in another PR. The purpose of this PR was reintegrating the same functionality of the original PR. Do you agree?

mbohlool replied:
I added an issue for that. I think we should CRDs need to register themselves with aggregator api server. Whatever we decide, though, I think it should be in another PR. The purpose of this PR was reintegrating the same functionality of the original PR. Do you agree?

sttts replied:
It's a different issue I think. The solution to #49878 might also solve the issue here though. It's the same thing as with the swagger comment. The swagger route has to query all go-restful containers in all delegates on the chain.

sttts replied:
But a follow-up is fine. This is broken in master and 1.7 as well. We have to fix this for 1.8.

mbohlool committed:
Bugfix: verify-no-vendor-cycles.sh detects wrong cycles

mbohlool commented:
I think tests should pass this time. @sttts as we discussed in slack, I self LGTM'ed this. Hopefully it will go through without a need for another rebase. I would move the discussion for your latest comments into the related issues and we can fix them in another PR.

mbohlool commented:
/test pull-kubernetes-e2e-gce-etcd3

mbohlool commented:
/test pull-kubernetes-e2e-gce-etcd3
